File: javascript/continuation_cps/fib_cps.js
// =============================================================================
// continuation_cps / fib_cps.js
//
// Concurrency Paradigm: Continuation-Passing Style (CPS)
//
// CPS transforms recursive functions by passing a "continuation" function
// that receives the result instead of returning it directly.  This makes
// control flow explicit and enables features like early exit, non-local
// return, and trampolining.
//
// Compare with: Python continuation_cps.py, Haskell ContMonad
// =============================================================================

// ---------------------------------------------------------------------------
// 1. Direct-style Fibonacci
// ---------------------------------------------------------------------------

function fibDirect(n) {
  if (n <= 1) return n;
  return fibDirect(n - 1) + fibDirect(n - 2);
}

// ---------------------------------------------------------------------------
// 2. CPS Fibonacci
// ---------------------------------------------------------------------------

function fibCps(n, cont) {
  if (n <= 1) return cont(n);
  return fibCps(n - 1, (a) => fibCps(n - 2, (b) => cont(a + b)));
}

// ---------------------------------------------------------------------------
// 3. CPS Factorial
// ---------------------------------------------------------------------------

function factCps(n, cont) {
  if (n === 0) return cont(1);
  return factCps(n - 1, (res) => cont(n * res));
}

// ---------------------------------------------------------------------------
// 4. Trampoline — avoids stack overflow for deep CPS
// ---------------------------------------------------------------------------

class Bounce {
  static call(thunk) {
    return { type: "call", thunk };
  }
  static done(value) {
    return { type: "done", value };
  }
}

function trampoline(b) {
  let result = b;
  while (result.type === "call") {
    result = result.thunk();
  }
  return result.value;
}

function fibCpsTramp(n, cont) {
  if (n <= 1) return cont(n);
  return Bounce.call(() =>
    fibCpsTramp(n - 1, (a) =>
      Bounce.call(() => fibCpsTramp(n - 2, (b) => cont(a + b)))
    )
  );
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

console.log("=== Continuation / CPS Demo ===\n");

// CPS Fibonacci
console.log("--- CPS Fibonacci ---");
for (let n = 0; n <= 10; n++) {
  const direct = fibDirect(n);
  const cps = fibCps(n, (x) => x);
  console.log(`  fib(${n}) = ${direct} (CPS: ${cps})`);
  console.assert(direct === cps, "Mismatch!");
}
console.log();

// CPS Factorial
console.log("--- CPS Factorial ---");
for (let n = 0; n <= 7; n++) {
  const cps = factCps(n, (x) => x);
  console.log(`  fact(${n}) = ${cps}`);
}
console.log();

// Trampoline (deep recursion without stack overflow)
console.log("--- Trampolined CPS ---");
const result = trampoline(fibCpsTramp(20, (x) => Bounce.done(x)));
console.log(`  fib_cps_trampoline(20) = ${result}`);
console.assert(result === fibDirect(20), "Mismatch!");
console.log();

console.log("=== Key insight ===");
console.log("CPS makes control flow explicit by passing continuations.");
console.log("Each recursive call receives a function describing");
console.log("'what to do next' instead of returning a value.");

File: lua/continuation_cps/fib_cps.lua
--[[
  fib_cps.lua
  Continuation-Passing Style (CPS) Fibonacci.

  Instead of returning values, each function calls a continuation
  function with the result. This makes control flow explicit.
  Compare with: Python continuation_cps.py, Haskell ContMonad
]]

-- Direct-style Fibonacci
local function fib_direct(n)
  if n <= 1 then return n end
  return fib_direct(n - 1) + fib_direct(n - 2)
end

-- CPS Fibonacci
local function fib_cps(n, cont)
  if n <= 1 then
    return cont(n)
  end
  return fib_cps(n - 1, function(a)
    return fib_cps(n - 2, function(b)
      return cont(a + b)
    end)
  end)
end

-- CPS Factorial
local function fact_cps(n, cont)
  if n == 0 then
    return cont(1)
  end
  return fact_cps(n - 1, function(res)
    return cont(n * res)
  end)
end

-- Tail-Recursive Trampoline (avoids stack overflow)
local Trampoline = {}
Trampoline.__index = Trampoline

function Trampoline.call(thunk)
  return setmetatable({ type = "call", thunk = thunk }, Trampoline)
end

function Trampoline.done(value)
  return setmetatable({ type = "done", value = value }, Trampoline)
end

function trampoline(b)
  local result = b
  while result.type == "call" do
    result = result.thunk()
  end
  return result.value
end

local function fib_cps_tramp(n, cont)
  if n <= 1 then
    return cont(n)
  end
  return Trampoline.call(function()
    return fib_cps_tramp(n - 1, function(a)
      return Trampoline.call(function()
        return fib_cps_tramp(n - 2, function(b)
          return cont(a + b)
        end)
      end)
    end)
  end)
end

-- Demo
print("=== Continuation / CPS Demo (Lua) ===\n")

-- CPS Fibonacci
print("--- CPS Fibonacci ---")
for n = 0, 10 do
  local direct = fib_direct(n)
  local cps = fib_cps(n, function(x) return x end)
  print(string.format("  fib(%d) = %d (CPS: %d)", n, direct, cps))
  assert(direct == cps)
end
print()

-- CPS Factorial
print("--- CPS Factorial ---")
for n = 0, 7 do
  local cps = fact_cps(n, function(x) return x end)
  print(string.format("  fact(%d) = %d", n, cps))
end
print()

-- Trampoline
print("--- Trampolined CPS ---")
local result = trampoline(fib_cps_tramp(20, function(x) return Trampoline.done(x) end))
print(string.format("  fib_cps_trampoline(20) = %d", result))
print(string.format("  matches direct: %s", tostring(result == fib_direct(20))))
print()

print("Key insight: CPS makes control flow explicit.")
print("Each function takes a continuation describing 'what to do next'.")
